File: src/main/java/core/procedures/strings/StringProc.java
package core.procedures.strings;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMMutableString;

public final class StringProc extends AFn {

  public StringProc() {
    super(new FnArgsBuilder().rest(Character.class).build());
  }

  @Override
  public String getName() {
    return "string";
  }

  @Override
  public SCMMutableString apply(Object... args) {
    if (args.length == 0) {
      return new SCMMutableString();
    }
    SCMMutableString string = new SCMMutableString(args.length);
    for (Object c : args) {
      string.append(c);
    }
    return string;
  }
}

File: src/main/java/core/procedures/strings/StringProc.kt
package core.procedures.strings

import core.procedures.AFn
import core.scm.MutableString

class StringProc : AFn<Any?, MutableString>(name = "string", isPure = true, restArgsType = Char::class.javaObjectType) {

    override operator fun invoke(args: Array<out Any?>) = args.joinTo(MutableString(), separator = "")
}
